tangelo/toolboxes/circuits/lcu.py

```diff
@@ -0,0 +1,310 @@
+    vector = np.sqrt(vector) / np.sqrt(vsum)
+    n_qubits = math.ceil(math.log2(num_terms))
+    newvec = np.zeros(2 ** n_qubits)
+    newvec[:num_terms] = vector
+
+    # Calculate circuit that generates vector and apply controls for Taylor series
+    s = StateVector(newvec, order="msq_first")
+    qss = list()
+    for k in range(kmax):
+        qss.append(s.initializing_circuit())
+        # Reindex to correct block of qubits for kth order of Taylor series
+        qss[-1].reindex_qubits(list(range(kmax + k * n_qubits, kmax + (k + 1) * n_qubits)))
+        # Add control for the kth unary encoded qubits
+        for gate in qss[-1]._gates:
+            if gate.control is not None:
+                gate.control += [k]
+            else:
+                gate.name = "C"+gate.name
+                gate.control = [k]
+    qstot = Circuit()
+    for qs in qss:
+        qstot += qs
+
+    # Calculate coefficients for unary encoding k="0"+"0"*(kmax-k)*"1"*k where
+    # first qubit encodes extra identity term needed to ensure 1-norm = 2 such that
+    # oblivious amplitude amplification is applicable
+    kvec = np.zeros(2**(kmax + 1))
+    for k in range(0, kmax+1):
+        pos = int("0" + "0" * (kmax - k) + "1" * k, base=2)
+        kvec[pos] = (t / time_steps * vsum) ** k / math.factorial(k)
+    kvec[0] += ((2 - expvsum) / 2)
+    kvec[int("1" + "0" * kmax, base=2)] = ((2 - expvsum) / 2)
+    kvec = np.sqrt(kvec) / np.sqrt(np.sum(kvec))
+    ksv = StateVector(kvec, order="msq_first")
+    kprep = ksv.initializing_circuit()
+
+    # shift encoding of extra identity term to last qubit
+    kprep.reindex_qubits(list(range(kmax)) + [kmax + kmax * n_qubits])
+
+    return kprep + qstot, unitaries, time_steps
+
+
+def USelectkl(unitaries: List[QubitOperator], n_qubits_sv: int, kmax: int, control: Union[int, List[int]] = None) -> Circuit:
+    r"""Generate the truncated Taylor series U_{Select} circuit for the list of QubitOperator as defined arXiv:1412.4687
+    The returned Circuit will have qubits defined in registers |n_qubits_sv>|kmax-1>|n_qubits_u>^{kmax-1}|ancilla>
+    n_qubits_sv is the number of qubits to define the state to propagate. |kmax-1> defines the unary encoding
+    of the Taylor series order. kmax-1 copies of length log2(len(unitaries)) to define the binary encoding of the linear
+    combination of unitaries. Finally, one ancilla qubit to ensure the success of oblivious amplitude amplification.
+    Args:
+        unitaries (list[QubitOperator]): The list of unitaries that defines the U_{Select} operation
+        n_qubits_sv (int): The number of qubits in the statevector register
+        kmax (int): The maximum Taylor series order
+        control (int or list[int]): Control qubits for operation
+    Returns:
+        Circuit: The circuit that implements the truncated Taylor series U_{Select} operation
+    """
+
+    n_qubits_u = math.ceil(math.log2(len(unitaries)))
```

Comment:
do we need a check on the resulting value, can `len(unitaries)` be anything or should it be a power of 2?

Reply:
The `math.ceil` rounds up to the nearest qubit number greater than `math.log2(len(unitaries))` so the function takes care of the power of 2.